Feat learning units precondition

Elscha opened the pull request:
Pull request to facilitate review

AhmadAlamoush committed:
feat(learning units precondition): Change required skills for learning units to skills expression

Add flexibility to select a skill expression as required skills e.g. ((Skill1 or skill2) and skill3)

AhmadAlamoush committed:
feat(learning units precondition): Add more tests

Add more tests to increase the coverage

AhmadAlamoush committed:
feat(learning units precondition): use ExpressionValues class

Use ExpressionValues class type to hold the skills and nested SkillExpression

AhmadAlamoush committed:
feat(learning units precondition): use Variable class

Try to simplify the usage of precondition for end users by passing only SkillExpression (as a Variable object) to the AND & OR.

AhmadAlamoush committed:
feat(learning units precondition): use Variable class

Fixing issues

AhmadAlamoush committed:
feat(learning units precondition): Add without skills for a skill expression

Add flexibility to create a skill expression for a list of skills with one skill e.g. (Loops Without For Each)

AhmadAlamoush committed:
feat(learning units precondition): Add an empty skill expression

- An empty skill expression for learning units without required skills
- Fixing errors

Elscha commented on src/lib/ast/and.ts:
```diff
@@ -0,0 +1,51 @@
+		return createJson(And.name, this.terms);
+	}
+
+	filterSkillsByWithout(skillsRelations: SkillsRelations, without?: Variable[]): SkillExpression[] {
```
I'm concerned that this function won't work recursively. That happens in the following situation:
`OR(AND(A, B, C), AND(D, E, F)), WITHOUT (A, E)`
Will be there an exception for `A` and `E` when I evaluate the `OR`-expression?

AhmadAlamoush replied:
The recursion is achieved through the evaluate function, which passes the "without" list into the nested expressions for evaluation. 
Note that in this example, the "Without" list is ignored since the skills are explicitly listed. However, let's assume that is not the case.

<img width="498" alt="NestedSkillExpressionWithout" src="https://github.com/e-Learning-by-SSE/nm-skill-lib/assets/75734718/37f561ec-130c-47e1-9ff0-b9c3f07e571a">

Elscha replied:
Looks good (convincing) 👍

Elscha commented on src/lib/ast/and.ts:
```diff
@@ -0,0 +1,51 @@
+import { Variable } from "./variable";
+import { SkillsRelations } from "./skillsRelation";
+
+// And skill expression
```
Please use JSDoc with meaning full comments, for instance as follows:
```typescript
/*
 * Conjunctive (AND) expression of which all terms must be fulfilled.
 * @param terms ...
 * @author Alamoush
 */
```

Elscha commented on src/lib/ast/empty.ts:
```diff
@@ -0,0 +1,31 @@
+import { SkillsRelations } from "./skillsRelation";
+
+// And skill expression
+export class Empty extends SkillExpression {
```
- For what is the empty expression needed?
- Comment was copied from `AND` class.

AhmadAlamoush replied:
An empty skill expression used in case the learning skill does not have any requirement.
The evaluation of this skill expression is always true.
We use this to ensure consistency across all learning units.

Elscha replied:
Wouldn't it be easier to make a preconditions in general **optional**? This way we need to check that a condition is present, but we do not need to create empty objects that need to be evaluated.

Elscha commented on src/lib/ast/jsonHandler.ts:
```diff
@@ -0,0 +1,57 @@
+
+// Convert a SkillExpression to a Json format
+export function createJson(operator: string, terms: SkillExpression[]): string {			
+    let nestedVariable: string[] = [];
```
Function seems to work as expected (recursively), but the naming of this variable seems not to fit as in general this does not store `nestedVariables`.
I would suggest to name it like `translatedTerms`, `jsonTerms`, `jsonifiedTerms`, ...

Elscha commented on src/lib/ast/jsonHandler.ts:
```diff
@@ -0,0 +1,57 @@
+    return jsonExpression;
+}
+
+// Parse a Json format to a SkillExpression
```
Please add JSDoc, for what is skillRelations needed?

Elscha commented on src/lib/ast/jsonHandler.ts:
```diff
@@ -0,0 +1,57 @@
+// Parse a Json format to a SkillExpression
+export function parseJsonExpression(json: string, skillsRelations: SkillsRelations): SkillExpression {	
+
+    const skill = skillsRelations.skills.find(sk => json == sk.id);
```
For the moment (testing) OK, but in future this may become very inperformant if in the first moment all strings are tested if they are a Skill and this function also uses a loop.

For instance: `OR(AND(A, B, C), AND(D, E, F)), WITHOUT (A, E)` the following terms should be checked if they are a skill:
- `OR(AND(A, B, C), AND(D, E, F))`
- `AND(A, B, C)`
- `AND(D, E, F)`

AhmadAlamoush replied:
The translation from JSON to SkillExpression is recursively. The stopping criterion is reaching a Skill. We could not know that unless we found a Skill in the skills list. There is no naming pattern for the skill id to check.
For example:
Since ("**OR(AND(A, B, C), AND(D, E, F))**") is not a Skill, then recursive call for **AND(A, B, C)** & **AND(D, E, F)**
Since ("**AND(A, B, C)**") is not a Skill, then recursive call for **A, B, & C**
Since ("**AND(D, E, F)**") is not a Skill, then recursive call for **D, E, & F**
Since **A** is a Skill return Variable of **A**
...
Last create skill expression for each set of Variable

Elscha replied:
I see that this works, but I was concerned about the runtime since you iterate over all skills for each string to check and you check also all AND and OR expressions.

Since you also store also the operator/type, wouldn't it be beneficial to search for Skill IDs only in case of `operator = variable`?

AhmadAlamoush replied:
At this point, we did not convert JSON string to Skill Expression yet (no operator yet, next line is the parsing). 
But, we have the JSON string and we could use it.

I changed it to check the JSON string.
if (!json.includes("operator")) {
const skill = skillsRelations.skills.find(sk => json == sk.id);
...
}

Elscha commented on src/lib/ast/jsonHandler.ts:
```diff
@@ -0,0 +1,57 @@
+        skills: skillExpression.skills
+    }
+
+    const variables: SkillExpression[] = [];
```
Naming

Elscha commented on src/lib/ast/or.ts:
```diff
@@ -0,0 +1,50 @@
+		return createJson(Or.name, this.terms);
+	}
+
+	filterSkillsByWithout(skillsRelations: SkillsRelations, without?: Variable[]): SkillExpression[] {
```
Not recusrive

AhmadAlamoush replied:
Which part is not recursive?

Elscha replied:
You convinced me in an other comment -> marked as resolved.

Elscha commented on src/lib/ast/skillExpression.ts:
```diff
@@ -0,0 +1,25 @@
+// extractSkills: return all skill in the expression as array of Skills
+// toJson: convert the expression into Json format
+
+export abstract class SkillExpression {
```
Please use JSDoc

Elscha commented on src/lib/ast/skillExpresssion.spec.ts:
```diff
@@ -0,0 +1,280 @@
+
+	const skillsRelations = new SkillsRelations(skills);
+
+	it("skillExpression type", () => {
```
Wrong indentation, please use "Prettier - Code formatter" addon for VS Code.

Elscha commented on src/lib/ast/skillExpresssion.spec.ts:
```diff
@@ -0,0 +1,280 @@
+		expect(emptySkills.toJson()).toBe("");
+	});
+
+	it("skills in precondition formula", () => {
```
Usually, tests should be as atomic as possible -> Test only one case per function, aggregate similar tests to one suite (describe), use meaning ful naming.

I would recommend a structure like this (may also organized in other hierarchies):
```typescript
describe("precondition formula", () => {
    ...
    describe("AND", () => {
    ...
        it("Top-Level skills; No nested expressions; fulfilled -> evaluates true)", () => {
            ...
        }
        it("Top-Level skills; No nested expressions; not fulfilled -> evaluates false)", () => {
            ...
        }
        it("Nested skills; No nested expressions; fulfilled -> evaluates true)", () => {
            ...
        }
        ...
        it("Top-Level skills; Nested expressions; fulfilled -> evluates true)", () => {
            ...
        }
        ...
    }
    describe("OR", () => {
        ...
    }
}
```

Elscha commented on src/lib/ast/skillsRelation.ts:
```diff
@@ -0,0 +1,19 @@
+		this.parents = skills.filter(skill => skill.nestedSkills.length > 0);
+	}
+
+	getChildren(parent: Skill): Skill[] {
```
Seems to have a poor performance as you do not use the `parents`property of the class. Instead you always iterate over all skills and check if they are the parent of the skill to be searched for...

The idea of such a class is to reduce the amount of loop-calls while evaluating all the expressions.

Elscha commented on src/lib/fastDownward/fastDownward.ts:
```diff
@@ -3,20 +3,27 @@ import { SearchNode } from "./searchNode";
 	const usefulLus = learningUnits
 		.filter(unit =>
-			unit.requiredSkills.every(skill => currentState.learnedSkills.includes(skill.id))
+			unit.requiredSkills.evaluate(currentState.learnedSkills, skillsRelations)
```
Does this consider `without` expressions?

AhmadAlamoush replied:
Yes, since the "without" is passed as a paramter to the evaluate function.

Elscha replied:
I dont see the parameter here

AhmadAlamoush replied:
The "without" parameter is optional to the evaluate function (we did not use it yet).

Elscha commented on src/lib/fastDownward/fastDownward.ts:
```diff
@@ -41,7 +48,7 @@ export function computeCost<LU extends LearningUnit>(
 			? // Check if the current LU requires any skills that are provided by the LU of the currentNode, only if a penalty is defined
 			  lu.teachingGoals.some(
-					skill => currentNode.action?.requiredSkills.includes(skill) ?? true
+					skill => currentNode.action?.requiredSkills.extractSkills().includes(skill) ?? true
```
Very good point, that I forgott in our talks.

But I ask myself how often variables are extracted from expressions at this point and if it could be avoided (by chachig the results per expression)...

Elscha commented on src/lib/fastDownward/fastDownward.ts:
```diff
@@ -245,7 +252,7 @@ export function search<LU extends LearningUnit>(
 				const tempLU = {
 					id: "-1",
-					requiredSkills: remainSkills,
+					requiredSkills: new And(remainSkills.map(skill => new Variable(skill))),
```
I'm unsure. Is this realy the new intention, that we always create an `AND` expression at this point?

AhmadAlamoush replied:
I think that it does not matter which skill expression I use in this step ('Or', 'And', or 'Empty'), since I am just passing the missing skill (or the unreachable skill).

Elscha reviewed (commented):
How do the changes affect the overall performance. Have you conducted any tests?

AhmadAlamoush committed:
feat(learning units precondition): solved part of the comments

- I solved part of the review comments.
- I added my comments on 5 review comments.

AhmadAlamoush commented:
> How do the changes affect the overall performance. Have you conducted any tests?

Yes, I tested the old use case that we had.
{
  "goal": ["2504"],
  "optimalSolution": true
}

The precondition increases the running time by approximately 1.5 seconds. But, when I do a quick test with the new merged approach (filtering the learning units "inScopeLearningUnits") it took around 35ms with the precondition

AhmadAlamoush committed:
feat(learning units precondition): Solving the comments on the pull

I solved some reviews and commented on the rest.

Elscha committed:
perf(SkillExpressions): SkillRelations performance fix

Creates SkillRelations only once instead of each iteration during availableOperations. Saves about 17% runtime (1.629 ms -> 1.363 ms).
Also fixed null checks in test cases

Elscha committed:
perf(SkillExpressions): ParentSkill  map instead of array

Performance boost about 40%

Elscha committed:
fix(SkillExpressions): SkillRelation support to return parents array

Elscha committed:
rf: Barrel exports for SkillExpressions

Elscha committed:
Merged main branch
